Add additional information to docs for serialization (#460)
bufbuild/protobuf-es: docs/runtime_api.md

User.typeName; // docs.User
```

## Serializing messages

> All messages provide functions for serializing and parsing data between the 

Reviewer: ```suggestion
All messages provide methods for serializing and parsing data between the 
```

User.fromBinary(protoBase64.dec(base64));
```

### `toBinary`

> The `toBinary` function accepts an options object which can be used to 

Reviewer: ```suggestion
The `toBinary` method accepts an options object which can be used to 
```

- `writeUnknownFields?: boolean`<br/>
   By default, unknown fields are included in the serialized output. This option
   allows for overriding that behavior.
   For more details see https://protobuf.dev/programming-guides/proto3/#unknowns.
- `writerFactory?: () => IBinaryWriter`<br/>
  A function for specifying a custom implementation to encode binary data.

For serializing multiple messages of the same type, see [size-delimited messages](#size-delimited-messages).

### `fromBinary`

The `fromBinary` function also accepts an options object which can be used to 
customize the serialization behavior. Supported options are:

- `readUnknownFields?: boolean`<br/>
   By default, unknown fields are retained during parsing and included in the 
   serialized output. This option allows for overriding this behavior.
   For more details see https://protobuf.dev/programming-guides/proto3/#unknowns.
- `readerFactory?: (bytes: Uint8Array) => IBinaryReader`<br/>
  A function for specifying a custom implementation to decode binary data.

##

Serializing to and from JSON can be done in a similar fashion:

```typescript
const json = user.toJson();
User.fromJson(json);
```

### `toJson`

> The `toJson` function also accepts an options object which can be used to 

Reviewer: ```suggestion
The `toJson` method also accepts an options object which can be used to 
```

   enum value instead.
- `useProtoFieldName?: boolean`<br/>
   By default the proto3 JSON printer should convert the field name to 
   lowerCamelCase and use that as the JSON name. An implementation may provide 
   an option to use the proto field name as the JSON name instead. Proto3 JSON 
   parsers are required to accept both the converted lowerCamelCase name and the 
   proto field name.
- `typeRegistry?: IMessageTypeRegistry`<br/>
   A type registry to use when parsing. This is required to write 
   `google.protobuf.Any` to JSON format.

### `toJsonString`

Note that the result of `toJson` will be a [JSON value][src-json-value] – a 
primitive JavaScript object that can be converted to a JSON string with the 
built-in function `JSON.stringify()`. For convenience, we also provide a method 
`toJsonString` that includes the stringify step:

```typescript
// Equivalent to:
// const json = user.toJson();
// const jsonStr = JSON.stringify(json);

const jsonStr = user.toJsonString();
```

Since `toJsonString` calls `toJson` under the hood, it accepts the same options 
object as `toJson` as well as additional options for the stringify step:

> - `emitDefaultValues?: boolean`<br/>

Reviewer: Shouldn't that be "prettySpaces"?

Author: Ugh yes, fixed.

>    A convenience property for the space option to `JSON.stringify`.
>    It indicates the number of space characters to be used as indentation, clamped to 10 
>    (that is, any number greater than 10 is treated as if it were 10). 
>    Values less than 1 indicate that no space should be used.
>    For more information, see the [`JSON.stringify` docs](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify).

Reviewer: Very wordy. WDYT about:

```suggestion
   A convenience property for the `space` option to `JSON.stringify`, which controls indentation for prettier output. 
   See the [`JSON.stringify` docs](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify#parameters).
```

> Note that messages also contain a protected `toJSON` function, which is invoked
> when `JSON.stringify` is used. For more information, see the [MDN docs](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify#tojson_behavior).
> However, you should never need to invoke this function, which is why
> it is marked as protected. 

Reviewer: My first question would be _why_ we have a toJSON, and what it does. What do you think about:

```suggestion
Note that all message have a `toJSON` method as well. If you serialize a message with `JSON.stringify()`, it will automatically call this method. By providing `toJSON`, we make sure that messages are always serialized to the well-defined protobuf JSON format. Otherwise, `JSON.stringify()` would crash on BigInt values, and would not serialize `oneof`, enumerations, and other types correctly. Serializing a message with `JSON.stringify()` is equivalent to `message.toJsonString({emitDefaultValues: true})`.
```

Author: Sure. There's two additional points that I think we should make though:

- By providing `toJSON`, we make sure that messages are always serialized to the well-defined protobuf JSON format _because it calls our `toJson` implementation_. I don't want to give the impression that just because we declared the method, we automatically get well-defined protobuf JSON.
- Users shouldn't need to use this method directly.

I augmented what you have with these two points. Let me know what you think.

As a rule of thumb, use `JSON.stringify` or `toJsonString` for stringified JSON. 
If actual JSON is desired, use `toJson`.

### `fromJson`

> The `fromJson` function also accepts an options object for customizing the 

Reviewer: ```suggestion
The `fromJson` method also accepts an options object for customizing the 
```
